test/systemtest/utils.py

```diff
@@ -0,0 +1,428 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+import difflib
+import filecmp
+import inspect
+import locale
+import logging
+import os
+import re
+import select
+import shlex
+import signal
+import subprocess
+import time
+from pathlib import Path
+
+import serial
+
+log = logging.getLogger(__name__)
+
+
+class Process:
+    """Utility class used to spawn an interact with processes."""
+    def __init__(self,
+                 cmd,
+                 logdir,
+                 cwd=None,
+                 startup_done_expect=None,
+                 startup_timeout=None,
+                 default_filter_func=None):
+        """Creates Process instance.
+
+        Args:
+            cmd: Command line argument used to spawn process.
+            logdir: Directory used to store STDOUT and STDERR output.
+            cwd: Working directory used to spawn process.
+            startup_done_expect: Pattern used to check at process startup time.
+            startup_timeout: Timeout in seconds for |startup_done_expect| check.
+            default_filter_func: Default function to filter all stdout/stderr
+                output with when matching it with find_in_output().
+        """
+        if isinstance(cmd, str):
+            self.cmd = shlex.split(cmd)
+        else:
+            self.cmd = [str(c) for c in cmd]
+        self.logdir = str(logdir)
+        self.cwd = str(cwd)
+        self.startup_done_expect = startup_done_expect
+        self.startup_timeout = startup_timeout
+        self.proc = None
+        self.logger = logging.getLogger(__name__)
+
+        self._f_stdout = None
+        self._f_stderr = None
+        self._f_stdout_r = None
+        self._f_stderr_r = None
+
+        self.default_filter_func = default_filter_func
+
+    def __del__(self):
+        try:
+            self.proc.kill()
+            self._f_stdout.close()
+            self._f_stderr.close()
+            self._f_stdout_r.close()
+            self._f_stderr_r.close()
+        except:
+            pass
+
+    def run(self):
+        """Start process with command line configured in constructor."""
+        cmd_name = os.path.basename(self.cmd[0])
+
+        # Enforce line-buffered STDOUT even when sending STDOUT/STDERR to file.
+        # If applications don't fflush() STDOUT manually, STDOUT goes through
+        # a 4kB buffer before we see any output, which prevents searching for
+        # the string indicating a successful startup.
+        # see discussion at http://www.pixelbeat.org/programming/stdio_buffering/
+        cmd = ['stdbuf', '-oL'] + self.cmd
```

Comment:
I'm not a massive fan of LD_PRELOAD hacks. Do we have any programs that actually need this? Otherwise, I'd be tempted to avoid it, and fix problems as they come up.

Reply:
Ok, understood.